### PR TITLE
Feature/cart list order

### PR DESCRIPTION
**Pull request by barshathakuri**
depends on https://github.com/toggle-corp/kitab-server/pull/15
Add order page and cart page

### CODE REVIEW

File: app/Base/configs/routes.tsx [hunk 1]
```diff
@@ -58,13 +58,33 @@ const bookDetail = wrap({
     },
     visibility: 'is-anything',
 });
+const cartPage = wrap({
+    path: '/cart/',
+    title: 'My Cart Page',
```
Comment (AdityaKhatri) [inline app/Base/configs/routes.tsx]:
```suggestion
    title: 'My Cart',
```

File: app/Base/configs/routes.tsx [hunk 2]
```diff
@@ -58,13 +58,33 @@ const bookDetail = wrap({
...
+    navbarVisibility: true,
+    component: lazy(() => import('#views/CartPage')),
+    componentProps: {
+    },
+    visibility: 'is-authenticated',
+});
+const orderPage = wrap({
+    path: '/order/',
+    title: 'My Order',
```
Comment (AdityaKhatri) [inline app/Base/configs/routes.tsx]:
```suggestion
    title: 'My Orders',
```

File: app/views/CartPage/index.tsx [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    }
+`;
+
+const DELETE_CART_ITEMS = gql`
+    mutation DeleteCartItems ($id: ID!) {
+        deleteCartItem(id: $id) {
+        errors
+        ok
+        }
+    }
```
Comment (AdityaKhatri) [inline app/views/CartPage/index.tsx]: use proper indentation

File: app/Base/components/Navbar/index.tsx [hunk 4]
```diff
@@ -134,6 +137,12 @@ function Navbar(props: Props) {
                     >
                         Login
                     </SmartButtonLikeLink>
+                    <ButtonLikeLink
+                        variant="secondary"
+                        to="/cart"
```
Comment (tnagorra) [inline app/Base/components/Navbar/index.tsx]: Please do not use link directly.
Use link from routes file

File: app/Base/configs/routes.tsx (hunk 1 above)
Comment (tnagorra) [inline app/Base/configs/routes.tsx]:
```suggestion
    title: 'My Cart',
```

File: app/views/BookDetail/index.tsx [hunk 5]
```diff
@@ -107,15 +108,18 @@ function BookDetail() {
                                 )}
                                 footerIcons={(
                                     <>
-                                        <Button name="buy">
+                                        <ButtonLikeLink
+                                            variant="primary"
+                                            to={`/order?book=${result.book.id}`}
```
Comment (tnagorra) [inline app/views/BookDetail/index.tsx]: Let's not prepare link directly. Please use path from routes page

File: app/views/BookDetail/index.tsx [hunk 6]
```diff
@@ -107,15 +108,18 @@ function BookDetail() {
...
+                                        >
                                             Buy now
-                                        </Button>
-                                        <Button
-                                            name="wishlist"
-                                            variant="secondary"
+                                        </ButtonLikeLink>
+                                        <ButtonLikeLink
+                                            variant="primary"
+                                            to="/wishlist-page"
```
Comment (tnagorra) [inline app/views/BookDetail/index.tsx]: Let's not prepare link directly. Please use path from routes page

File: app/views/OrderPage/index.tsx [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+            errors
+            ok
+        result {
+            orderCode
+            id
+            status
+            totalPrice
+            }
+        }
+    }
```
Comment (tnagorra) [inline app/views/OrderPage/index.tsx]: The indentation for gql is not correct

File: app/views/OrderPage/index.tsx [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+
+function OrderItem(props: ItemProps) {
+    const { book, quantity, changeQuantity } = props;
+    const { id, title, price } = book;
+    const handleQuantityChange = (value: number | undefined,
+        name: string,
+        event: React.FormEvent<HTMLInputElement> | undefined) => {
+        if (value) changeQuantity(value);
+        else changeQuantity(1);
+    };
```
Comment (tnagorra) [inline app/views/OrderPage/index.tsx]: Please use useCallback

File: app/views/OrderPage/index.tsx [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+function OrderItem(props: ItemProps) {
+    const { book, quantity, changeQuantity } = props;
+    const { id, title, price } = book;
+    const handleQuantityChange = (value: number | undefined,
+        name: string,
+        event: React.FormEvent<HTMLInputElement> | undefined) => {
+        if (value) changeQuantity(value);
+        else changeQuantity(1);
```
Comment (tnagorra) [inline app/views/OrderPage/index.tsx]: Let's not skip curly braces.

Also, it's better to just use ternary operator here

File: app/views/OrderPage/index.tsx [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+                        />
+                        <TextOutput
+                            label="Price (NPR)"
+                            valueType="number"
+                            value={price}
+                        />
+                        <TextOutput
+                            label="Amount (NPR)"
+                            valueType="number"
+                            value={price * quantity}
```
Comment (tnagorra) [inline app/views/OrderPage/index.tsx]: Is price an quantity always defined?

File: app/views/OrderPage/index.tsx [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+                        </div>
+                        <TextOutput
+                            label="Quantity"
+                            valueType="number"
+                        />
+                        <NumberInput
+                            name="quantity"
+                            value={quantity}
+                            onChange={handleQuantityChange}
+                        />
```
Comment (tnagorra) [inline app/views/OrderPage/index.tsx]: Why do we have output and input for quantity?

File: app/views/OrderPage/index.tsx [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+                </Container>
+                <div className={styles.orderButton}>
+                    <Button
+                        name={undefined}
+                        onClick={undefined}
+                        variant="secondary"
+                        icons={<BsCashStack />}
+                    >
+                        Cash On Delivery
+                    </Button>
```
Comment (tnagorra) [inline app/views/OrderPage/index.tsx]: Please disable this button if not implemented

File: app/views/OrderPage/index.tsx [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+                </div>
+            </div>
+        </div>
+    );
+}
+
+function OrderPage() {
+    const location = useLocation();
+    const params = new URLSearchParams(location.search);
+    const id = params.get('book');
```
Comment (tnagorra) [inline app/views/OrderPage/index.tsx]: Please use hook provided by react router to get params

File: app/views/OrderPage/index.tsx [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+    const [quantity, setQuantity] = useState<number>(1);
+    const { data: result, loading } = useQuery<
+        BookByIdQuery,
+        BookByIdQueryVariables
+    >(BOOK_DETAIL, {
+        skip: !id,
+        variables: { id: id ?? '' },
```
Comment (tnagorra) [inline app/views/OrderPage/index.tsx]: Let's just define variables as undefined when id is not defined

File: app/views/OrderPage/index.tsx [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+            SingleOrderMutationVariables
+        >(ORDER);
+    const submit = useCallback(() => {
+        const bookId = id ? parseInt(id, 10) : 0;
+        placeSingleOrder({ variables: { bookId, qty: quantity } });
+    }, [id, quantity]);
+
+    const setQuantityChange = (qty: number) => {
+        setQuantity(qty);
+    };
```
Comment (tnagorra) [inline app/views/OrderPage/index.tsx]: Any reason to define this function? We can just use setQuantity

File: app/views/CartPage/index.tsx [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+            {!loading && carts.length > 0
+                && (
+                    <Button
+                        name={undefined}
+                        variant="secondary"
+                        onClick={checkout}
+                    >
+                        Order
+                    </Button>
+                )}
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx]: Let's only disable the button when loading

File: app/views/CartPage/index.tsx [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+
+    const checkout = () => {
+        placeOrderFromCart().then(() => {
+            refetch();
+        }).catch((e) => {
+            console.log(e);
+        });
+    };
+
+    const cartItemRendererParams = React.useCallback((_, data) => ({
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx]: Add typings for data

File: app/views/CartPage/index.tsx [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+            refetch();
+        });
+    }, []);
+
+    const checkout = () => {
+        placeOrderFromCart().then(() => {
+            refetch();
+        }).catch((e) => {
+            console.log(e);
+        });
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx]: Any reason  to use promise instead of onCompleted?

File: app/views/CartPage/index.tsx [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+            CheckOutCartMutationVariables
+        >(ORDER_FROM_CART);
+
+    const pending = loading || submitting;
+    const carts = (!loading && result?.cartItems?.results) ? result.cartItems.results : [];
+
+    const removeCartItem = useCallback((id: string) => {
+        deleteCartItem({ variables: { id } }).then(() => {
+            refetch();
+        });
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx]: Any reason to use promise instead of onCompleted?

File: app/views/CartPage/index.tsx [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+    });
+
+    const [placeOrderFromCart,
+        { data: response, loading: submitting }] = useMutation<
+            CheckOutCartMutation,
+            CheckOutCartMutationVariables
+        >(ORDER_FROM_CART);
+
+    const pending = loading || submitting;
+    const carts = (!loading && result?.cartItems?.results) ? result.cartItems.results : [];
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx]: We don't need to check loading state to read results

File: app/views/CartPage/index.tsx [hunk 21]
```diff
@@ -0,0 +1,238 @@
...
+                </div>
+            </div>
+        </div>
+    );
+}
+
+function CartPage() {
+    const [page, setPage] = useState<number>(1);
+    const [pageSize, setPageSize] = useState<number>(10);
+    const [email, setEmail] = useState<string>('admin@gmail.com');
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx]: Let's remove default email

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: app/Base/components/Navbar/index.tsx [hunk 22]
```diff
@@ -134,6 +137,12 @@ function Navbar(props: Props) {
                     >
                         Login
                     </SmartButtonLikeLink>
+                    <ButtonLikeLink
```
Comment (AdityaKhatri) [inline app/Base/components/Navbar/index.tsx]: Use SmartButtonLikeLink

File: app/views/CartPage/index.tsx [hunk 23]
```diff
@@ -0,0 +1,238 @@
+import React, { useCallback, useState } from 'react';
+import { Button, Container, ListView, Message, NumberInput, TextOutput } from '@the-deep/deep-ui';
```
Comment (AdityaKhatri) [inline app/views/CartPage/index.tsx]: spread in multiple lines

File: app/views/CartPage/index.tsx [hunk 24]
```diff
@@ -0,0 +1,238 @@
...
+import {
+    gql,
+    useMutation,
+    useQuery,
+} from '@apollo/client';
+import { AiTwotoneDelete } from 'react-icons/ai';
+import { FaHeart } from 'react-icons/fa';
```
Comment (AdityaKhatri) [inline app/views/CartPage/index.tsx]: If possible use io5, if not in io5 use io4. Under drastic circumstances use other libraries

File: app/views/CartPage/index.tsx [hunk 25]
```diff
@@ -0,0 +1,238 @@
...
+    } = props;
+
+    const {
+        id,
+        quantity,
+        book,
+    } = cart;
+
+    const {
+        title, image,
```
Comment (AdityaKhatri) [inline app/views/CartPage/index.tsx:104]: drop in multiple lines

File: app/views/CartPage/index.tsx [hunk 26]
```diff
@@ -0,0 +1,238 @@
...
+        CartListQueryVariables
+    >(CART_LIST, {
+        variables: { page, pageSize, email },
+    });
+
+    const [placeOrderFromCart,
+        { data: response, loading: submitting }] = useMutation<
+            CheckOutCartMutation,
+            CheckOutCartMutationVariables
+        >(ORDER_FROM_CART);
```
Comment (AdityaKhatri) [inline app/views/CartPage/index.tsx]:
```suggestion
    const [
        placeOrderFromCart,
        {
            data: response,
            loading: submitting,
        },
    ] = useMutation<CheckOutCartMutation, CheckOutCartMutationVariables>(
        ORDER_FROM_CART,
        {
            onCompleted: () => {
            },
        },
    );

```

File: app/views/CartPage/index.tsx (hunk 17 above)
Comment (AdityaKhatri) [inline app/views/CartPage/index.tsx]:
```suggestion
    const cartItemRendererParams: CartContentProps = React.useCallback((_, data: Cart) => ({
```

File: app/views/CartPage/index.tsx [hunk 27]
```diff
@@ -0,0 +1,238 @@
...
+            refetch();
+        }).catch((e) => {
+            console.log(e);
+        });
+    };
+
+    const cartItemRendererParams = React.useCallback((_, data) => ({
+        cart: data,
+        deleteCart: removeCartItem,
+    }), []);
```
Comment (AdityaKhatri) [inline app/views/CartPage/index.tsx]:
```suggestion
    }), [removeCartItem]);
```

File: app/views/CartPage/index.tsx [hunk 28]
```diff
@@ -0,0 +1,238 @@
...
+                    <Button
+                        name={undefined}
+                        variant="secondary"
+                        onClick={checkout}
+                    >
+                        Order
+                    </Button>
+                )}
+        </div>
+
```
Comment (AdityaKhatri) [inline app/views/CartPage/index.tsx]: remove empty line

File: app/views/OrderPage/index.tsx [hunk 29]
```diff
@@ -0,0 +1,181 @@
...
+    quantity: number;
+    changeQuantity: (qty: number) => void;
+}
+
+function OrderItem(props: ItemProps) {
+    const { book, quantity, changeQuantity } = props;
+    const { id, title, price } = book;
+    const handleQuantityChange = (value: number | undefined,
+        name: string,
+        event: React.FormEvent<HTMLInputElement> | undefined) => {
```
Comment (AdityaKhatri) [inline app/views/OrderPage/index.tsx]: remove unused arguments from function

File: app/views/OrderPage/index.tsx (hunk 10 above)
Comment (AdityaKhatri) [reply app/views/OrderPage/index.tsx]: and*

File: app/views/OrderPage/index.tsx (hunk 14 above)
Comment (AdityaKhatri) [inline app/views/OrderPage/index.tsx]:
```suggestion
        variables: id ? { id } : undefined,
```

File: app/views/OrderPage/index.tsx [hunk 30]
```diff
@@ -0,0 +1,181 @@
...
+                            Order #
+                            {resp.placeSingleOrder.result?.orderCode}
+                            For Total Price: NPR
+                            {resp.placeSingleOrder.result?.totalPrice}
+                            Order Status:
+                            {resp.placeSingleOrder.result?.status}
+                        </Alert>
+                    </>
+
+                )}
```
Comment (AdityaKhatri) [inline app/views/OrderPage/index.tsx]: Remove for now, we'll need to setup alerts separately

File: app/views/OrderPage/index.tsx [hunk 31]
```diff
@@ -0,0 +1,181 @@
...
+            {
+                result && result.book && !loading
```
Comment (AdityaKhatri) [inline app/views/OrderPage/index.tsx]:
```suggestion
                result?.book && !loading
```

**Review by AdityaKhatri — CHANGES_REQUESTED** [review]

File: app/views/BookDetail/index.tsx (hunk 6 above)
Comment (barshathakuri) [reply app/views/BookDetail/index.tsx]: this is fixed in PR #16

File: app/views/OrderPage/index.tsx (hunk 13 above)
Comment (barshathakuri) [reply app/views/OrderPage/index.tsx]: we need to upgrade the react-router-dom to version 6. In this, we have version 5.

File: app/views/OrderPage/index.tsx (hunk 30 above)
Comment (barshathakuri) [reply app/views/OrderPage/index.tsx]: TODO

File: app/views/CartPage/index.tsx [hunk 32]
```diff
@@ -0,0 +1,251 @@
...
+        }
+        quantity
+      }
+      grandTotalPrice
+      page
+      pageSize
+      totalCount
+    }
+  }
+`;
```
Comment (AdityaKhatri) [inline app/views/CartPage/index.tsx]: use 4 space indentation

File: app/views/BookDetail/index.tsx [hunk 33]
```diff
@@ -20,28 +20,30 @@ import {
...
 const BOOK_DETAIL = gql`
-query BookDetail ($id: ID!){
-    book(id: $id ) {
-        description
+query BookDetail($id: ID!) {
+    book(id: $id) {
+      description
+      id
+      image {
+        name
```
Comment (tnagorra) [inline app/views/BookDetail/index.tsx:37]: Let's fetch id if image has id

File: app/views/CartPage/index.tsx [hunk 34]
```diff
@@ -0,0 +1,265 @@
...
+    cartItems(createdBy: $email, page: $page, pageSize: $pageSize) {
+      results {
+        id
+        totalPrice
+        book {
+          id
+          title
+          image {
+            url
+          }
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx]: Let's fetch id if image has id

File: app/views/OrderPage/index.tsx [hunk 35]
```diff
@@ -0,0 +1,206 @@
...
+                        />
+                        <TextOutput
+                            label="Price (NPR)"
+                            valueType="number"
+                            value={price}
+                        />
+                        <TextOutput
+                            label="Amount (NPR)"
+                            valueType="number"
+                            value={isDefined(quantity) ? price * quantity : 0}
```
Comment (tnagorra) [inline app/views/OrderPage/index.tsx]: isn't quantity always defined?

File: app/views/OrderPage/index.tsx [hunk 36]
```diff
@@ -0,0 +1,206 @@
...
+                </Container>
+                <div className={styles.orderButton}>
+                    <Button
+                        name={undefined}
+                        onClick={undefined}
+                        variant="secondary"
+                        icons={<IoCashOutline />}
+                    >
+                        Cash On Delivery
+                    </Button>
```
Comment (tnagorra) [inline app/views/OrderPage/index.tsx]: Let's disable the button if it's not implemented

File: app/views/OrderPage/index.tsx (hunk 13 above)
Comment (tnagorra) [reply app/views/OrderPage/index.tsx]: We don't need to use version 6.

Why are we using search params btw?

File: app/views/OrderPage/index.tsx [hunk 37]
```diff
@@ -0,0 +1,206 @@
...
+    });
+
+    const alert = useAlert();
+
+    const [placeSingleOrder,
+        {
+            data: resp,
+            loading: submitting,
+        },
+    ] = useMutation<
```
Comment (tnagorra) [inline app/views/OrderPage/index.tsx]: Let's move placeSingleOrder to next line

File: app/views/OrderPage/index.tsx [hunk 38]
```diff
@@ -0,0 +1,206 @@
...
+                        `Order #
+                        ${resp?.placeSingleOrder?.result?.orderCode}
+                        For Total Price: NPR
+                        ${resp?.placeSingleOrder?.result?.totalPrice}
+                        Order Status:
+                        ${resp?.placeSingleOrder?.result?.status}`,
+                        {
+                            variant: 'success',
+                        },
+                    );
```
Comment (tnagorra) [inline app/views/OrderPage/index.tsx]: This may result in "undefined" text on alert.
We should avoid passing undefined values to formatted string

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: app/views/BookDetail/index.tsx (hunk 33 above)
Comment (barshathakuri) [reply app/views/BookDetail/index.tsx:37]: id not available for image in server

**Review by AdityaKhatri — APPROVED** [review]

**Pull request merged**